### PR TITLE
WIP for developing greylisting

### PR DESCRIPTION
**Pull request by jamescowens**
Some initial thoughts. Just a skeleton mostly... I really don't like the cs_main lock in the whitelist snapshot, but I want to be able to add the greylist status to the snapshot, which requires getting the current SB...

### CODE REVIEW

**Comment by jamescowens** [comment]
I am thinking about how we fill out the tally class for WAS.... We have to have current SB index - 40's TC, current SB - 7's TC and current SB's TC...

**Comment by jamescowens** [comment]
The way I have handled the greylisting in the scraper is that it treats greylisted projects as whitelisted, EXCEPT for two things... one, it does not include the greylisted project in the nActiveProjects count to determine the per project magnitude, and two, it only populates the project level statistic entry (no CPID entries) for the greylisted projects. The project level entry only contains TC. The other statistics are zeroed out.

**Comment by jamescowens** [comment]
This approach essentially means the greylist is a true filter operator on the statististics files/objects to form the SB.

**Comment by cyrossignol** [comment]
I added a commit to the base `zcd` branch with a set of classes for obtaining the greylist status of projects in the whitelist that does not hamper the thread-safe design of the Whitelist API. I did not implement the rules themselves.

The goal is for it to work something like this: 

```c++ 
// First get a whitelist snapshot as usual. We may already have one:
const NN::WhitelistSnapshot projects = NN::GetWhitelist().Snapshot();

// Run each of the whitelisted projects through the greylist rules: 
//
// Depending on how we implement the greylist state, we may need to lock a 
// mutex to filter whitelists. A GreylistSnapshot/WhitelistSnapshot object
// is const, so we cannot initialize these by assignment in nested scopes. 
// We may move locks within the function to make these calls prettier, but
// this reduces synchronization to one place in the greylist API.
//
ENTER_CRITICAL_SECTION(cs_main);
const NN::GreylistSnapshot greylist = NN::Quorum::FilterGreylist(projects);
LEAVE_CRITICAL_SECTION(cs_main);

// Detemine greylist status of a particular project:
const GreylistReason reason = greylist.Check("Einstein");
if (greylist.CheckActive("Einstein")) { ... }
if (greylist.CheckGreylisted("Einstein")) { ... }

// Get number of active/greylisted projects:
const size_t nActiveProjects = greylist.ActiveProjects().size();
const size_t nGreylistedProjects = greylist.GreylistedProjects().size(); 

// Do something with all active projects:
for (const auto& active_project : greylist.ActiveProjects()) {
    const std::string project_name = active_project->m_name;
    DoSomething(project_name, *active_project);
}

// Do something with all greylisted projects:
for (const auto& greylisted_project : greylist.GreylistedProjects()) { 
    const std::string project_name = greylisted_project->m_name;
    const GreylistReason reason = greylisted_project.m_reason;
    DoSomething(project_name, *active_project, greylisted_reason);
}
```

By decoupling the greylist state from the whitelist API and filtering at the point-of-use, we reduce the interactions between contract updates and superblocks that may both refer to the same projects. Let me know what you think.

**Comment by jamescowens** [comment]
I like it.

File: src/neuralnet/greylist.h [hunk 1]
```diff
@@ -13,6 +13,8 @@ enum class GreylistReason
     ZERO_CREDIT_DAYS,  //!< Exported too many recent zero-credit days.
 };
 
+typedef std::vector<GreylistReason> GreylistReasons; //!< There can be more than one reason.
```
Comment (cyrossignol) [inline src/neuralnet/greylist.h:15]: Might be interesting to make a bitfield for `GreylistReason`s and define each reason as one bit position in that field. Then we can just check `== 0` for `GreylistReason::NONE`. Maybe for later... perhaps for now an empty `GreylistReasons` can represent `NONE` so we don't need to visit each reason to determine that.

File: src/scraper/scraper.cpp [hunk 2]
```diff
@@ -2776,7 +2778,23 @@ bool LoadProjectObjectToStatsByCPID(const std::string& project, const CSerialize
...
+    // the CPID level statistics.
+
+    const NN::WhitelistSnapshot projectWhitelist = NN::GetWhitelist().Snapshot();
+
+    const NN::GreylistSnapshot greylist = NN::Quorum::FilterGreylist(projectWhitelist);
+
+    // If the specified project is not on the whitelist, immmediately return false.
+    if (!projectWhitelist.Contains(project)) return false;
+
+    bool greylisted = greylist.Contains(project);
```
Comment (cyrossignol) [inline src/scraper/scraper.cpp]: `GreylistSnapshot::Contains()` just forwards to `WhitelistSnapshot::Contains()` for code that just needs to check whether the project existed when the snapshot was taken. I should have made the name more clear. To check greylist status, use `GreylistSnapshot::CheckActive()` or `GreylistSnapshot::CheckGreylisted()`.

Here, we want to use: 
```c++
bool greylisted = greylist.CheckGreylisted(project);
```

File: src/scraper/scraper.cpp [hunk 3]
```diff
@@ -2943,16 +2976,30 @@ ScraperStats GetScraperStatsByConsensusBeaconList()
...
         if (fDebug3) _log(logattribute::INFO, "LOCK", "get scraper stats by consensus beacon list - count active projects: cs_StructScraperFileManifest");
 
         for (auto const& entry : StructScraperFileManifest.mScraperFileManifest)
         {
-            //
-            if (entry.second.current && !entry.second.excludefromcsmanifest) nActiveProjects++;
+            // This checks to see if the project file entry is contained in the whitelist (it should be) AND is it not greylisted AND it is current
+            // AND is not excluded from CSManifest. The last is to handle explorer mode operation.
+            if (projectWhitelist.Contains(entry.first)
+                    && !greylist.Contains(entry.first)
```
Comment (cyrossignol) [inline src/scraper/scraper.cpp]: Here we want to use:
```c++
&& greylist.CheckActive(entry.first)
```

File: src/scraper/scraper.cpp [hunk 4]
```diff
@@ -3001,14 +3048,38 @@ ScraperStats GetScraperStatsByConsensusBeaconList()
...
+
+    const NN::WhitelistSnapshot projectWhitelist = NN::GetWhitelist().Snapshot();
+
+    const NN::GreylistSnapshot greylist = NN::Quorum::FilterGreylist(projectWhitelist);
+
+    // Note we cannot just use greylist.ActiveProjects().size() here, because the Manifest may be missing a
+    // project that is on the whitelist due to lack of availability.
+    for (const auto& entry : StructConvergedManifest.ConvergedManifestPartsMap)
+    {
+        if (projectWhitelist.Contains(entry.first) && !greylist.Contains(entry.first)) ++nActiveProjects;
```
Comment (cyrossignol) [inline src/scraper/scraper.cpp]: Here we want to use:

```c++
&& greylist.CheckActive(entry.first)
```

File: src/neuralnet/greylist.h (hunk 1 above)
Comment (jamescowens) [reply src/neuralnet/greylist.h:15]: I thought about that too. I don't know if the juice is worth the squeeze to do a bitmap style approach for this. Certainly in any case, eliminating the GreylistReason::NONE and requiring that none is implictly defined by the absence of an element in the vector of reasons drastically simplifies the filter to basically checking whether reasons.size() >0.

File: src/scraper/scraper.cpp (hunk 2 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: You're right. I even recoded the filter when I changed the reasons to a vector. I must just be tired. A general state these days unfortunately.

File: src/scraper/scraper.cpp (hunk 3 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: Same here.

File: src/scraper/scraper.cpp (hunk 4 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: Yep

Comment (jamescowens) [reply src/scraper/scraper.cpp]: And we don't need the Whitelist.Contains here.... it is superfluous.

**Pull request closed without merging**